### PR TITLE
[Organizations] Allow collaborators to create API keys scoped to an organization with the ability to push new IDs

### PR DESCRIPTION
**Pull request by scottbommarito**
Unfortunately missed this bug when fixing https://github.com/NuGet/NuGetGallery/issues/5731

### CODE REVIEW

File: src/NuGetGallery/Controllers/UsersController.cs [hunk 1]
```diff
@@ -855,29 +854,57 @@ public virtual ActionResult PasswordChanged()
             return credentialViewModel;
         }
 
-        // todo: integrate verification logic into PermissionsService.
-        private bool VerifyScopes(User scopeOwner, IEnumerable<Scope> scopes)
+        private static IDictionary<string, IActionRequiringEntityPermissions> AllowedActionToActionRequiringEntityPermissions = new Dictionary<string, IActionRequiringEntityPermissions>
```
Comment (loic-sharma) [inline src/NuGetGallery/Controllers/UsersController.cs]: Nit: fields normally are defined before methods.

Comment (scottbommarito) [reply src/NuGetGallery/Controllers/UsersController.cs]: I thought this was fine given that it's exclusively used for this method haha

Comment (loic-sharma) [reply src/NuGetGallery/Controllers/UsersController.cs]: Also, would the name `AllowedScopeToActionRequiringEntityPermissions`?

Comment (scottbommarito) [reply src/NuGetGallery/Controllers/UsersController.cs]: No--unfortunately `NuGetScopes` was named poorly and has a misleading name. `NuGetScopes` makes a lot more sense if you think of it as `NuGetScopeActions`, which is why I named this `AllowedActionToActionRequiringEntityPermissions`.

File: src/NuGetGallery/Controllers/UsersController.cs [hunk 2]
```diff
@@ -855,29 +854,57 @@ public virtual ActionResult PasswordChanged()
...
-                if (organization != null)
+                if (string.IsNullOrEmpty(scope.AllowedAction))
+                {
+                    // All scopes must have an allowed action.
+                    return false;
+                }
+                
+                // Get the list of actions allowed by this scope.
+                var actions = new List<IActionRequiringEntityPermissions>();
+                foreach (var allowedAction in AllowedActionToActionRequiringEntityPermissions.Keys)
```
Comment (chenriksson) [inline src/NuGetGallery/Controllers/UsersController.cs]: Why three `for` loops? Isn't there 1 `ActionsRequiringPermissions` per `NuGetScopes`?

File: src/NuGetGallery/Controllers/UsersController.cs [hunk 3]
```diff
@@ -855,29 +854,57 @@ public virtual ActionResult PasswordChanged()
...
         {
-            var currentUser = GetCurrentUser();
-
-            // scoped to the user
-            if (currentUser.MatchesUser(scopeOwner))
+            { NuGetScopes.PackagePush, ActionsRequiringPermissions.UploadNewPackageId },
+            { NuGetScopes.PackagePushVersion, ActionsRequiringPermissions.UploadNewPackageVersion },
+            { NuGetScopes.PackageUnlist, ActionsRequiringPermissions.UnlistOrRelistPackage },
+            { NuGetScopes.PackageVerify, ActionsRequiringPermissions.VerifyPackage },
+        };
```
Comment (chenriksson) [inline src/NuGetGallery/Controllers/UsersController.cs]: Can you just create a helper on `NuGetScopes` or `ActionsRequiringPermissions` that does the 1:1 conversion?

File: src/NuGetGallery/Controllers/UsersController.cs (hunk 2 above)
Comment (scottbommarito) [reply src/NuGetGallery/Controllers/UsersController.cs]: As far as I can tell from the code, the `AllowedAction` could be `NuGetScopes.All` which means we need to check the upload new, upload existing, unlist, and verify actions.

Comment (scottbommarito) [reply src/NuGetGallery/Controllers/UsersController.cs]: To clarify, the loop is best explained like this:

For each scope:
 1) Get the actions that it grants the user to perform
 2) Check that each action can be performed by the user

File: src/NuGetGallery/Controllers/UsersController.cs (hunk 3 above)
Comment (scottbommarito) [reply src/NuGetGallery/Controllers/UsersController.cs]: I was going to do this if `ApiScopeEvaluator` did any conversion similar to this, but it doesn't seem to be the case. This code doesn't seem to be useful to share outside of this file.

File: src/NuGetGallery/Controllers/UsersController.cs [hunk 4]
```diff
@@ -855,29 +854,57 @@ public virtual ActionResult PasswordChanged()
             return credentialViewModel;
         }
 
-        // todo: integrate verification logic into PermissionsService.
-        private bool VerifyScopes(User scopeOwner, IEnumerable<Scope> scopes)
+        private static IDictionary<string, IActionRequiringEntityPermissions[]> AllowedActionToActionRequiringEntityPermissions = new Dictionary<string, IActionRequiringEntityPermissions[]>
```
Comment (skofman1) [inline src/NuGetGallery/Controllers/UsersController.cs]: perhaps call this field something with "map"?

**Pull request merged**